community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/BigEndianByteArrayBuffer.java

```diff
@@ -0,0 +1,296 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.store.kvstore;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+import org.neo4j.io.pagecache.PageCursor;
+
+final class BigEndianByteArrayBuffer implements ReadableBuffer, WritableBuffer
+{
+    static BigEndianByteArrayBuffer buffer( int size )
+    {
+        return new BigEndianByteArrayBuffer( size );
+    }
+
+    final byte[] buffer;
+
+    BigEndianByteArrayBuffer( int size )
+    {
+        this( new byte[size] );
+    }
+
+    BigEndianByteArrayBuffer( byte[] buffer )
+    {
+        this.buffer = buffer;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toString( buffer );
+    }
+
+    static String toString( byte[] buffer )
+    {
+        StringBuilder result = new StringBuilder( buffer.length * 6 + 1 ).append( '[' );
+        for ( byte b : buffer )
+        {
+            if ( b >= 32 && b < 127 )
+            {
+                if ( b == '\'' )
+                {
+                    result.append( "'\\''" );
+                }
+                else
+                {
+                    result.append( '\'' ).append( (char) b ).append( '\'' );
+                }
+            }
+            else
+            {
+                result.append( "0x" );
+                if ( b < 16 )
+                {
+                    result.append( 0 );
+                }
+                result.append( Integer.toHexString( b & 0xFF ) );
+            }
+            result.append( ", " );
+        }
+        if ( result.length() > 1 )
+        {
+            result.setLength( result.length() - 2 );
+        }
+        result.append( ']' );
+        return result.toString();
+    }
+
+    static int compare( byte[] key, byte[] searchSpace, int offset )
+    {
+        for ( int i = 0; i < key.length; i++ )
+        {
+            int result = (key[i] & 0xFF) - (searchSpace[offset + i] & 0xFF);
+            if ( result != 0 )
+            {
+                return result;
+            }
+        }
+        return 0;
+    }
+
+    public void clear()
+    {
+        fill( (byte) 0 );
+    }
+
+    public void fill( byte zero )
+    {
+        Arrays.fill( buffer, zero );
+    }
+
+    public boolean allZeroes()
+    {
+        for ( byte b : buffer )
+        {
+            if ( b != 0 )
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public void dataFrom( ByteBuffer buffer )
+    {
+        buffer.get( this.buffer );
+    }
+
+    public void dataTo( byte[] target, int targetPos )
+    {
+        assert target.length >= targetPos + buffer.length : "insufficient space";
+        System.arraycopy( buffer, 0, target, targetPos, buffer.length );
+    }
+
+    public int size()
+    {
+        return buffer.length;
+    }
+
+    public byte getByte( int offset )
+    {
+        if ( offset > size() - 1 )
```

Comment:
sure, and maybe introduce a range-check method as you suggest in https://github.com/neo4j/neo4j/pull/3852#discussion_r23916060
with a nice message, as you suggest in https://github.com/neo4j/neo4j/pull/3852#discussion_r23916099